tools/compiletest/Cargo.toml

```diff
@@ -0,0 +1,14 @@
+[package]
+name = "compiletest"
+description = "Run tests using compiletest-rs"
+version = "0.0.0"
+edition = "2021"
+
+[dependencies]
+compiletest_rs = { version = "0.10.0", features = [ "rustc" ] }
```

Comment:
Ok... I'm finding `ui_test` a bit harder to use (probably because I'm used to compiletest). Do you have any example I could use?

I noticed that `run_tests()` will read all the arguments for the command, which doesn't let me configure any extra argument (like the path for the driver binary).

The other option would be using [`run_tests_generic`](https://docs.rs/ui_test/0.20.0/ui_test/fn.run_tests_generic.html#), but that seems more complicated than what I want.

I think a `run_tests_with_args(&args)` would be ideal.

Reply:
Actually, I'll just copy the body of `run_tests()` but change how we build args.

Reply:
Now I see a bunch of errors like:

```
error: a compiletest-rs style comment was detected.
Please use text that could not also be interpreted as a command,
and prefix all actual commands with `//@`
```

I'll push my changes here, but I might need help.

Reply:
@oli-obk can you please advise how to properly use `ui_test` here? Right now, the current script result is:
```
test result: FAIL. 26 failed; 11 passed;
```

And I get a bunch of failures like:
<pre><font color="#F66151"><b>error</b></font>: <b>no `:` after level found</b>
  <font color="#2A7BDE"><b>--&gt;</b></font> tests/ui/cfg/diagnostics-same-crate.rs:32:28
<font color="#2A7BDE"><b>   |</b></font>
<font color="#2A7BDE"><b>32 |</b></font>     inner::uwu(); //~ ERROR cannot find function
<font color="#2A7BDE"><b>   |</b></font><font color="#F66151"><b>                            ^^^^^^^^^^^^^^^^^^^^^</b></font>
<font color="#2A7BDE"><b>   |</b></font>

<font color="#F66151"><b>error</b></font>: <b>`//~|` pattern without preceding line</b>
  <font color="#2A7BDE"><b>--&gt;</b></font> tests/ui/cfg/diagnostics-same-crate.rs:33:8
<font color="#2A7BDE"><b>   |</b></font>
<font color="#2A7BDE"><b>33 |</b></font>     //~| NOTE not found in `inner`
<font color="#2A7BDE"><b>   |</b></font><font color="#F66151"><b>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^</b></font>
<font color="#2A7BDE"><b>   |</b></font>

</pre>

Reply:
oh... you're trying to run this against the rustc test suite? That's not going to work even with the compiletest-rs crate. We'd need to use the compiletest crate from within the rustc repo, it's the only thing that actually supports the various flags that the rustc ui test suite uses.

So you'd not need to implement anything for running the test suite, but instead run `cargo run --manifest-path rustc_dir/src/tools/compiletest/Cargo.toml --rustc-path=path_to_test_drive --src-base=rustc_dir/tests/ui cfg` plus presumably some other flags

Reply:
Compiletest-rs and ui_test need in-tree test suites. I was thinking we'd add some examples here. We could have test/demo drivers that each run on their own or a shared test suite and produce some output that we want to check to detect issues. One example could be testing various `Debug` printings of SMIR data structures

Reply:
Yes, I was thinking we could add stable_mir specific tests, and for that we could use the code here.

I think running the rustc or other pre-existing suite would be helpful to ensure we have good test coverage